src/main/java/rx/internal/operators/OperatorMerge.java

```diff
@@ -71,715 +72,878 @@
+            boolean success = false;
+            synchronized (this) {
+                // if nobody is emitting and child has available requests
+                if (!emitting && producer.get() > 0) {
+                    emitting = true;
+                    success = true;
+                }
+            }
+            if (success) {
+                boolean skipFinal = false;
+                try {
+                    try {
+                        child.onNext(value);
+                    } catch (Throwable t) {
+                        skipFinal = true;
+                        Exceptions.throwIfFatal(t);
+                        try {
+                            child.onError(t);
+                        } finally {
+                            unsubscribe();
+                        }
+                        return;
                     }
-                } else if (r > 0) {
-                    // drain what was requested
-                    long toEmit = r;
-                    for (int i = 0; i < toEmit; i++) {
-                        Object o = svq.poll();
-                        if (o == null) {
-                            break;
-                        } else {
-                            on.accept(actual, o);
-                            emittedWhileDraining++;
+                    producer.produced(1);
+                    request(1);
+                    // check if some state changed while emitting
+                    synchronized (this) {
+                        skipFinal = true;
+                        if (!missed) {
+                            emitting = false;
+                            return;
                         }
+                        missed = false;
                     }
-                    // decrement the number we emitted from outstanding requests
-                    MergeProducer.REQUESTED.getAndAdd(mergeProducer, -emittedWhileDraining);
-                }
-                return emittedWhileDraining;
-            }
-            return 0;
-        }
-
-        final Func1<InnerSubscriber<T>, Boolean> DRAIN_ACTION = new Func1<InnerSubscriber<T>, Boolean>() {
-
-            @Override
-            public Boolean call(InnerSubscriber<T> s) {
-                if (s.q != null) {
-                    long r = mergeProducer.requested;
-                    int emitted = s.drainQueue();
-                    if (emitted > 0) {
-                        s.requestMore(emitted);
+                } finally {
+                    if (!skipFinal) {
+                        synchronized (this) {
+                            emitting = false;
+                        }
                     }
-                    if (emitted == r) {
-                        // we emitted as many as were requested so stop the forEach loop
-                        return Boolean.FALSE;
+                }
+                // in case there was some activity, continue with the regular emission loop
+                emitLoop();
```

Comment:
It is part of the fast-path recovery here and is called if `emitting == true` and there was a missed true -> false transition just before it. Maybe the diff-view is too distracting in this case. Otherwise, I'm not sure what additional documentation it needs.